// react-native-curved-bottom-bar/lib/commonjs/CurvedBottomBar/components/navigator/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _bottomTabs = require("@react-navigation/bottom-tabs");

var _react = _interopRequireWildcard(require("react"));

var _reactNative = require("react-native");

var _reactNativeSvg = _interopRequireWildcard(require("react-native-svg"));

var _useDeviceOrientation = require("../../../useDeviceOrientation");

var _path = require("./path");

var _styles = require("./styles");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

const {
  width: maxW
} = _reactNative.Dimensions.get('window');

const Tab = (0, _bottomTabs.createBottomTabNavigator)();

const BottomBarComponent = /*#__PURE__*/_react.default.forwardRef((props, ref) => {
  const SVG = _reactNativeSvg.default;
  const PATH = _reactNativeSvg.Path;
  const {
    type = 'DOWN',
    style,
    width = null,
    height = 65,
    circleWidth = 50,
    bgColor = 'gray',
    initialRouteName,
    tabBar,
    renderCircle,
    borderTopLeftRight = false,
    strokeWidth = 0
  } = props;
  const [itemLeft, setItemLeft] = (0, _react.useState)([]);
  const [itemRight, setItemRight] = (0, _react.useState)([]);
  const [maxWidth, setMaxWidth] = (0, _react.useState)(width || maxW);
  const children = props === null || props === void 0 ? void 0 : props.children;
  const orientation = (0, _useDeviceOrientation.useDeviceOrientation)();
  (0, _react.useImperativeHandle)(ref, () => {
    return {};
  });
  (0, _react.useEffect)(() => {
    const {
      width: w
    } = _reactNative.Dimensions.get('window');

    if (!width) {
      setMaxWidth(w);
    }
  }, [orientation, width]);

  const _renderButtonCenter = navigate => {
    var _children$filter$, _children$filter$$pro;

    const getTab = (_children$filter$ = children.filter(e => {
      var _e$props;

      return (e === null || e === void 0 ? void 0 : (_e$props = e.props) === null || _e$props === void 0 ? void 0 : _e$props.position) === 'CENTER';
    })[0]) === null || _children$filter$ === void 0 ? void 0 : (_children$filter$$pro = _children$filter$.props) === null || _children$filter$$pro === void 0 ? void 0 : _children$filter$$pro.name;
    return renderCircle({
      selectedTab: getTab,
      navigate: selectTab => {
        if (selectTab) {
          navigate(selectTab);
        }
      }
    });
  };

  (0, _react.useEffect)(() => {
    const arrLeft = children.filter(item => {
      var _item$props;

      return (item === null || item === void 0 ? void 0 : (_item$props = item.props) === null || _item$props === void 0 ? void 0 : _item$props.position) === 'LEFT';
    });
    const arrRight = children.filter(item => {
      var _item$props2;

      return (item === null || item === void 0 ? void 0 : (_item$props2 = item.props) === null || _item$props2 === void 0 ? void 0 : _item$props2.position) === 'RIGHT';
    });
    setItemLeft(arrLeft);
    setItemRight(arrRight);
  }, [children, initialRouteName]);
  const d = type === 'DOWN' ? (0, _path.getPath)(maxWidth, height, circleWidth >= 50 ? circleWidth : 50, borderTopLeftRight) : (0, _path.getPathUp)(maxWidth, height + 30, circleWidth >= 50 ? circleWidth : 50, borderTopLeftRight);

  const renderItem = _ref => {
    let {
      color,
      routeName,
      navigate
    } = _ref;
    return /*#__PURE__*/_react.default.createElement(_reactNative.TouchableOpacity, {
      key: routeName,
      style: _styles.styles.itemTab,
      onPress: () => navigate(routeName)
    }, /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
      style: {
        color: color
      }
    }, routeName));
  };

  const MyTabBar = props => {
    const {
      state,
      navigation
    } = props;
    const focusedTab = state === null || state === void 0 ? void 0 : state.routes[state.index].name;
    return /*#__PURE__*/_react.default.createElement(_reactNative.View, null, /*#__PURE__*/_react.default.createElement(_reactNative.View, {
      style: [_styles.styles.container, style]
    }, /*#__PURE__*/_react.default.createElement(SVG, {
      width: maxWidth,
      height: height + (type === 'DOWN' ? 0 : 30)
    }, /*#__PURE__*/_react.default.createElement(PATH, {
      fill: bgColor,
      stroke: "#DDDDDD",
      strokeWidth: strokeWidth,
      d
    })), /*#__PURE__*/_react.default.createElement(_reactNative.View, {
      style: [_styles.styles.main, {
        width: maxWidth
      }, type === 'UP' && _styles.styles.top30]
    }, /*#__PURE__*/_react.default.createElement(_reactNative.View, {
      style: [_styles.styles.rowLeft, {
        height: height
      }]
    }, itemLeft.map((item, index) => {
      var _item$props3;

      const routeName = item === null || item === void 0 ? void 0 : (_item$props3 = item.props) === null || _item$props3 === void 0 ? void 0 : _item$props3.name;

      if (tabBar === undefined) {
        return renderItem({
          routeName,
          color: focusedTab === routeName ? 'blue' : 'gray',
          navigate: navigation.navigate
        });
      }

      return /*#__PURE__*/_react.default.createElement(_reactNative.View, {
        style: _styles.styles.flex1,
        key: index.toString()
      }, tabBar({
        routeName,
        selectedTab: focusedTab,
        navigate: selectTab => {
          if (selectTab !== focusedTab) {
            navigation.navigate({
              name: routeName,
              merge: true
            });
          }
        }
      }));
    })), _renderButtonCenter(navigation.navigate), /*#__PURE__*/_react.default.createElement(_reactNative.View, {
      style: [_styles.styles.rowRight, {
        height: height
      }]
    }, itemRight.map((item, index) => {
      var _item$props4;

      const routeName = item === null || item === void 0 ? void 0 : (_item$props4 = item.props) === null || _item$props4 === void 0 ? void 0 : _item$props4.name;

      if (tabBar === undefined) {
        return renderItem({
          routeName,
          color: focusedTab === routeName ? 'blue' : 'gray',
          navigate: navigation.navigate
        });
      }

      return /*#__PURE__*/_react.default.createElement(_reactNative.View, {
        style: _styles.styles.flex1,
        key: index.toString()
      }, tabBar({
        routeName,
        selectedTab: focusedTab,
        navigate: selectTab => {
          if (selectTab !== focusedTab) {
            navigation.navigate({
              name: routeName,
              merge: true
            });
          }
        }
      }));
    })))));
  };

  return /*#__PURE__*/_react.default.createElement(Tab.Navigator, _extends({}, props, {
    tabBar: MyTabBar
  }), children === null || children === void 0 ? void 0 : children.map((e, index) => {
    return /*#__PURE__*/_react.default.createElement(Tab.Screen, _extends({
      key: index.toString()
    }, e.props));
  }));
});

var _default = BottomBarComponent;
exports.default = _default;
